mask/eolearn/mask/cloud_mask.py

```diff
@@ -292,3 +297,563 @@ def get_s2_pixel_cloud_detector(threshold=0.4, average_over=4, dilation_size=2,
+                                         diff_interweaved
+                                         ),
+                                        axis=3
+                                        )
+
+        multi_features = multi_features.reshape(np.prod(multi_features.shape[:-1]), multi_features.shape[-1])
+
+        return multi_features
+
+    def execute(self, eopatch):
+        """
+        Add selected features (cloud probabilities and masks) to an EOPatch instance.
+
+        :param eopatch: Input `EOPatch` instance
+        :return: `EOPatch` with additional features
+        """
+
+        # Get data
+        bands = eopatch.data[self.data_feature][..., self.band_indices].astype(np.float32)
+        is_data = eopatch.mask[self.is_data_feature].astype(bool)
+
+        original_shape = bands.shape[1:-1]
+        scale_factors, self.sigma = self._parse_resolution_data(original_shape, eopatch.meta_info)
+
+        mono_proba_feature, mono_mask_feature = self.mono_features
+        multi_proba_feature, multi_mask_feature = self.multi_features
+
+        # Downscale if specified
+        if scale_factors is not None:
+            bands = resize_images(bands.astype(np.float32), scale_factors=scale_factors)
+            is_data = resize_images(is_data.astype(np.uint8), scale_factors=scale_factors).astype(np.bool)
+
+
+        mono_proba = None
+        multi_proba = None
+
+        # Run s2cloudless if needed
+        if any(feature is not None for feature in [self.mask_feature, mono_mask_feature, mono_proba_feature]):
+            mono_proba = self._mono_iterations(bands)
+            mono_proba = mono_proba.reshape(*bands.shape[:-1], 1)
+
+            # Upscale if necessary
+            if scale_factors is not None:
+                mono_proba = resize_images(mono_proba, new_size=original_shape)
+
+            # Average over and threshold
+            mono_mask = self._average_all(mono_proba) >= self.mono_threshold
+
+        # Run SSIM-based multi-temporal classifier if needed
+        if any(feature is not None for feature in [self.mask_feature, multi_mask_feature, multi_proba_feature]):
+            multi_proba = self._multi_iterations(bands, is_data)
+            multi_proba = multi_proba.reshape(*bands.shape[:-1], 1)
+
+            # Upscale if necessary
+            if scale_factors is not None:
+                multi_proba = resize_images(multi_proba, new_size=original_shape)
+
+            # Average over and threshold
+            multi_mask = self._average_all(multi_proba) >= self.multi_threshold
+
+
+        # Intersect
+        if self.mask_feature is not None:
+            inter_mask = mono_mask & multi_mask
+
+        # Add features
+        is_data = eopatch.mask[self.is_data_feature].astype(bool)
+
+        if mono_mask_feature is not None:
+            mono_mask = self._dilate_all(mono_mask)
+            eopatch.mask[mono_mask_feature] = (mono_mask * is_data).astype(bool)
```

Comment:
not sure if `is_data` should be included in the resulting masks and probabilities, or should be kept separate

Reply:
This step is only to ensure that masks stay within valid parts of the images. One can expect that areas, where `is_data == False`, are to be predicted as "clear" by the classifier itself, but dilation alone causes the masks to extend into invalid regions.

Reply:
I see. OK for me to include `is_data` here already